src/child-applications/lifecycles/unload.js

```diff
@@ -13,23 +13,34 @@ export async function toUnloadPromise(app) {
 		 * anything that still thinks we need to unload the app.
 		 */
 		finishUnloadingApp(app, unloadInfo);
-		return;
+		return app;
+	}
+
+	if (app.status === UNLOADING) {
```

Comment:
This fixes a race condition. Usually, we can assume that there is only one caller (the `reroute` function) of things like `toBootstrapPromise`, `toMountPromise`, etc. But in this case `toUnloadPromise`, it can be either `reroute` or it can be `immediatelyUnloadChildApp`. So we have to handle some concurrency here. It is possible that both code paths run while an app is being unloaded. This case is exposed by some of the tests, and in that case we want both callers to wait for the same promise.